### PR TITLE
Update UnitarySystem set point control

### LINKED ISSUE #10570: AirloopHVAC:UnitarySystem set point control does not turn on cooling coil
Issue overview
--------------
A user test file shows that the Coil:Cooling:WaterToAirHeatPump:VariableSpeedEquationFit does not turn on when using set point control. A refactor of the fan model may have inadvertently disabled setpoint control for UnitarySystem. the top right figure shows the set point temperature and outlet air temperature not tracking well. The lower left figure shows that the cooling coil does not turn on.

image

Bug description courtesy of Trane North America.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added ([UnitarySystem_VSHeatPumpWaterToAirEquationFit_SPControl.idf.txt](https://github.com/user-attachments/files/15891016/UnitarySystem_VSHeatPumpWaterToAirEquationFit_SPControl.idf.txt))

 - [ ] Ticket added to EnergyPlus Defect Complexity (Github Project)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #10570
 - Testing showed that set point controls in UnitarySystem did not turn on the cooling coil and other coils did not respond as anticipated.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by rraustad** [comment]
This branch shows all coils now turn on and set point is maintained except for when the AvailabilityManger is turning off the system at night for a few hours and on a weekend. I'm not exactly sure yet where the change in air flow from cooling to heating is coming from.

    Component Sizing Information, Fan:SystemModel, SUPPLY FAN 1, User-Specified Design Maximum Air Flow Rate [m3/s], 1.00000
    Component Sizing Information, AirLoopHVAC:UnitarySystem, WATER-TO-AIR HEAT PUMP UNIT 1, User-Specified Supply Air Flow Rate [m3/s], 1.00000
    Component Sizing Information, AirLoopHVAC:UnitarySystem, WATER-TO-AIR HEAT PUMP UNIT 1, User-Specified Heating Supply Air Flow Rate [m3/s], 1.00000
    Component Sizing Information, AirLoopHVAC:UnitarySystem, WATER-TO-AIR HEAT PUMP UNIT 1, User-Specified Cooling Supply Air Flow Rate [m3/s], 1.00000
    Component Sizing Information, AirLoopHVAC:UnitarySystem, WATER-TO-AIR HEAT PUMP UNIT 1, User-Specified No Load Supply Air Flow Rate [m3/s], 0.00000
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:VARIABLESPEEDEQUATIONFIT, SYS 1 HEAT PUMP COOLING MODE, Design Size Rated Air Flow Rate [m3/s], 1.00000
    Component Sizing Information, COIL:COOLING:WATERTOAIRHEATPUMP:VARIABLESPEEDEQUATIONFIT, SYS 1 HEAT PUMP COOLING MODE, User-Specified Rated Air Flow Rate [m3/s], 1.00000


image

**Comment by nrel-bot-3** [comment]
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 16 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 9 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 9 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 12 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 10 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 12 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 20 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 13 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 22 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@rraustad it has been 7 days since this pull request was last updated.

**Comment by mitchute** [comment]
@rraustad it looks like this one got lost. Are you still intending to move this PR forward?

**Comment by rraustad** [comment]
@mitchute I'll have to go back over this one. It doesn't look like much changed so I will just test again with the defect file to see if this branch activates the cooling coil while develop does not.

**Comment by rraustad** [comment]
The issue still exists in develop. There is no air flow rate through the UnitarySystem in develop and changes in this branch allow the UnitarySystem to turn the fan on.

develop result with defect file:

{908D8F0E-A5A7-4CF1-ADC4-7CE9B0FF6015}

this branch results:

{8597DC1F-5F78-439C-BA7C-1682FC60A1DF}

**Comment by rraustad** [comment]
@mitchute I looked at the diff file that had the biggest increase in warnings. I don't see any difference in operation so I don't have a good explanation for the change in number of warnings.

{9C340B06-4E9D-4BA5-B0C2-385AB7F3DD22}

This is a set point controlled UnitarySystem where the air flow rate is constant. There is no difference in air flow rate so a change in warnings may be due to a change in number of iterations, but that doesn't make much sense either since the air loop outlet node temperature and air flow rate did not change. I _think_ the change in warnings is unrelated but can't be sure until I know what causes these warnings, that were present before this code change.

image

**Comment by rraustad** [comment]
@mitchute I backed out all changes and started over. These are the minimum changes that get the system to at least turn on. I think I would rather have the system at least operate, albeit with some control issues, than not turn on at all. I will post a new issue to address the remaining issue which is out of scope and a user input control problem to either use the previous on air flow or off air flow (i.e. the m_AirFlowControl flag). This defect file is set point controlled which should honor this control choice.

    //  set air flow control mode:
    //  m_AirFlowControl = UseCompFlow::On means operate at last cooling or heating air flow requested when compressor is off
    //  m_AirFlowControl = UseCompFlow::Off means operate at no load air flow value specified by user
    //  AirFlowControl only valid if fan opmode = ContFanCycComp
    if (this->m_MaxNoCoolHeatAirVolFlow == 0.0) {
        this->m_AirFlowControl = UseCompFlow::On;
    } else {
        this->m_AirFlowControl = UseCompFlow::Off;
    }

**Comment by rraustad** [comment]
Results after these changes. Top right figure shows set point control is working.

{C2C9210C-9C0E-4FFD-81D9-F22759D29F14}


develop branch:

{74AC992F-49D8-45AE-A644-EAACEA0D85C3}

File: src/EnergyPlus/UnitarySystem.cc [hunk 1]
```diff
@@ -8234,6 +8235,11 @@ namespace UnitarySystems {
                 SingleDuct::SimATMixer(state, this->m_ATMixerName, FirstHVACIteration, this->m_ATMixerIndex);
             }
         }
+
+        // set variables used in fan call inside function calcPassiveSystem
+        state.dataUnitarySystems->m_massFlow1 = state.dataLoopNodes->Node(this->AirInNode).MassFlowRate;
+        state.dataUnitarySystems->m_runTimeFraction1 = (state.dataUnitarySystems->m_massFlow1 > 0.0) ? 1.0 : 0.0; // constant fan mode is required
+
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: If either of these changes are removed, the system air flow goes to 0 and there is no control to set point temperature. I'm not quite sure the initialization of LastMode should be done in getInput but I am warming up to it since the mode needs to start somewhere. If the mode is not set from 0 to cooling or heating mode the fan does not turn on. So this change seems OK. And if the fan operating parameters are not set, for set point controlled systems, then the fan won't turn on. So this also seems OK.

Comment (mitchute) [reply src/EnergyPlus/UnitarySystem.cc]: @rraustad do you think this is captured in our tests already? Given your description, it seems like we should have some sort of test to ensure this continues to function.

Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: This was hard to track down. The defect file has Fan:SystemModel in a modified example file (UnitarySystem_VSHeatPumpWaterToAirEquationFit). Unit tests have legacy fans. One unit test has Fan:System models but doesn't test controls. The Fan:SystemModel needs to set these 2 variables for the fan call in calcPassiveSystem. This may have started with the fan refactor or a change to UnitarySystem that did not notice this corner case. I'm trying to figure out a good place for a unit test.

Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: First 2 unit tests passed fine without any changes. Then figured out that it was the coil type and the fan type that caused this issue and a specific unit test was added. That unit test fails without these changes. This coil type was calling calcPassiveSystem when it did not need to. Those lines were deleted. Since that function is no longer called the new lines to set the fan parameters were also not needed.

**Review by mitchute — APPROVED** [review]
Looking good here. Merging.
